packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/ExtractionService.java

```diff
@@ -0,0 +1,288 @@
+package software.uncharted.terarium.hmiserver.service;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.http.HttpEntity;
+import org.apache.http.entity.ByteArrayEntity;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+import software.uncharted.terarium.hmiserver.models.ClientEvent;
+import software.uncharted.terarium.hmiserver.models.ClientEventType;
+import software.uncharted.terarium.hmiserver.models.dataservice.document.DocumentAsset;
+import software.uncharted.terarium.hmiserver.models.dataservice.document.DocumentExtraction;
+import software.uncharted.terarium.hmiserver.models.dataservice.document.ExtractionAssetType;
+import software.uncharted.terarium.hmiserver.models.extractionservice.ExtractionStatusUpdate;
+import software.uncharted.terarium.hmiserver.models.task.TaskRequest;
+import software.uncharted.terarium.hmiserver.proxies.documentservice.ExtractionProxy;
+import software.uncharted.terarium.hmiserver.service.data.DocumentAssetService;
+import software.uncharted.terarium.hmiserver.service.tasks.ModelCardResponseHandler;
+import software.uncharted.terarium.hmiserver.utils.ByteMultipartFile;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class ExtractionService {
+    final DocumentAssetService documentService;
+    final ExtractionProxy extractionProxy;
+    final ObjectMapper objectMapper;
+    final ClientEventService clientEventService;
+
+    private ExecutorService executor = Executors.newFixedThreadPool(1);
+
+    private static final Integer TOTAL_EXTRACTION_STEPS = 14;
+
+    public void extractPDF(UUID documentId, String userId) {
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    updateClient(documentId, 0, TOTAL_EXTRACTION_STEPS, userId);
+                    DocumentAsset document = documentService.getAsset(documentId).get();
+                    updateClient(documentId, 1, TOTAL_EXTRACTION_STEPS, userId);
+
+                    if (document.getFileNames().isEmpty()) {
+                        String errorMsg = "No files found on document";
+                        updateClientError(documentId, 1, TOTAL_EXTRACTION_STEPS, errorMsg, userId);
+                        throw new RuntimeException(errorMsg);
```

Comment:
Do we need to add this to the signature for the method?

Reply:
Actually I think `RuntimeException` doesn't need it, but, it may be worth it to add it or change the exception type here and below so that the controllers know to catch problems.

Reply:
This is in a Runnable - not the function extractPDF - instead the RuntimeException is to short cut the function allowing the Executor to free up the thread

Reply:
Ah thanks!